convenience methods for field embeddings

yyyyx4 opened the pull request:
Currently, the best(?) way to compute embeddings between fields is somewhat hidden inside the `Hom` object. In this patch we add helper methods to access such embeddings. In addition, we add an `.algebraic_closure()` method to non-prime finite fields.

These seem useful.

yyyyx4 committed:
convenience methods for field embeddings

yyyyx4 committed:
fix doctest

GiacomoPope commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -1871,6 +1871,12 @@ cdef class FiniteField(Field):
             z3
 
+        For finite fields, the algebraic closure is always (isomorphic
+        to) the algebraic closure of the prime field:
```
I think this needs a double colon for the block?

```suggestion
        to) the algebraic closure of the prime field::
```

yyyyx4 committed:
add missing colon

Co-authored-by: Giacomo Pope <44242839+GiacomoPope@users.noreply.github.com>

yyyyx4 committed:
support embeddings from non-Conway finite fields into the algebraic closure

yyyyx4 committed:
embedding from GF(p^k,'a') into GF(p).algebraic_closure() is not canonical

...hence the output here would be more or less meaningless. We should
fail instead, until algebraic closures of non-prime finite fields are
supported "properly".

(reverts part of 87e912ba168)

S17A05 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -1920,6 +1931,107 @@ cdef class FiniteField(Field):
+              From: Cyclotomic Field of order 3 and degree 2
+              To:   Cyclotomic Field of order 6 and degree 2
+              Defn: zeta3 |--> -zeta6
+            ]
```
I think these tests should not be here, since the fields to be embedded are not finite fields

S17A05 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -1920,6 +1931,107 @@ cdef class FiniteField(Field):
+        g = self.gen()
+        if (emb := K.coerce_map_from(self)) is not None:
+            return self.hom([emb(g)])
+        try:
+            r = g.minpoly().change_ring(K).any_root()
+        except ValueError:
+            raise ValueError(f'no embedding from {self} to {K}')
```
This can also raise a `TypeError` if I tested it correctly, e.g. when asking for an embedding from `GF(2)` into `GF(7).algebraic_closure()`; not sure if also excepting `TypeError` leads to some other issue, though

S17A05 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -1920,6 +1931,107 @@ cdef class FiniteField(Field):
+            pass
+        if K not in FiniteFields():
+            from sage.rings.algebraic_closure_finite_field import AlgebraicClosureFiniteField_generic
+            if not isinstance(K, AlgebraicClosureFiniteField_generic):
+                raise NotImplementedError('computing embeddings into this ring not implemented')
```
It might be a nice addition to compare here (or somewhere before) the characteristic of both fields, so that a lot of basic cases where embeddings cannot exist (e.g. `GF(2)` into `QQbar`) are caught with a more explicit error message.

S17A05 commented on src/sage/rings/ring.pyx:
```diff
@@ -1478,6 +1478,55 @@ cdef class Field(CommutativeRing):
+        except TypeError:
+            pass
+        from sage.categories.sets_cat import EmptySetError
+        try:
+            return H.an_element()
+        except EmptySetError:
+            raise ValueError(f'no embedding from {self} to {K}')
```
Unless I made a mistake in setup, this seems to not work as intended: E.g. for `GF(2).an_embedding(GF(7))` it does not raise an `EmptySetError`, and instead just returns `None`.

S17A05 commented on src/sage/rings/finite_rings/finite_field_base.pyx:
```diff
@@ -1920,6 +1931,107 @@ cdef class FiniteField(Field):
+        rs = []
+        if (emb := K.coerce_map_from(self)) is not None:
+            rs.append(emb(g))
+        rs += [r for r,_ in g.minpoly().roots(ring=K) if r not in rs]
+        return [self.hom([r]) for r in rs]
```
One more thing: This also seems to fail when considering different characteristics, e.g. `GF(2).embeddings(GF(7))` runs into the issue that it cannot defined a homorphism mapping `1` to `6`; possibly this and all the other issues can already be prevented by the characteristic comparison

yyyyx4 committed:
reviewer comments
